File: src/core/peerPins.js
// Trust-on-first-use pin store for remote identity keys.
//
// The first time a peer completes a signed (v3) handshake, we pin the SHA-256
// fingerprint of their identity public key. On every subsequent handshake we
// require the presented idPub to hash to the same fingerprint. A mismatch
// indicates one of:
//   1. The peer legitimately rotated keys (app reinstall, fresh identity).
//   2. An active MITM is substituting a different identity key.
//
// The application layer must surface (1) to the user — we cannot distinguish
// them automatically. This module deliberately does not auto-overwrite pins.

import { openDatabase } from './db.js';
import { bytesToBase64, base64ToBytes } from './base64.js';
import { computeFingerprint } from './identityKey.js';

const KEY_PREFIX = 'peer-pin-';

function rowKey(peerId) {
  return `${KEY_PREFIX}${String(peerId || '')}`;
}

/**
 * @returns {Promise<{ peerId: string, pubSpkiB64: string, fingerprint: string, pinnedAt: number } | null>}
 */
export async function getPin(peerId) {
  if (!peerId) return null;
  const db = await openDatabase();
  const row = await db.get('keys', rowKey(peerId));
  if (!row || !row.pubSpkiB64) return null;
  return {
    peerId: row.peerId || String(peerId),
    pubSpkiB64: String(row.pubSpkiB64),
    fingerprint: String(row.fingerprint || ''),
    pinnedAt: Number(row.pinnedAt || 0)
  };
}

export async function setPin(peerId, pubSpkiBytes) {
  if (!peerId || !pubSpkiBytes) throw new Error('setPin: peerId and pubSpkiBytes required');
  const fingerprint = await computeFingerprint(pubSpkiBytes);
  const db = await openDatabase();
  await db.put('keys', {
    id: rowKey(peerId),
    peerId: String(peerId),
    pubSpkiB64: bytesToBase64(pubSpkiBytes),
    fingerprint,
    pinnedAt: Date.now()
  });
  return { fingerprint };
}

export async function deletePin(peerId) {
  if (!peerId) return false;
  const db = await openDatabase();
  await db.delete('keys', rowKey(peerId));
  return true;
}

/**
 * Verify that a remote pubkey matches the existing pin for this peer.
 *
 * Returns one of:
 *   { status: 'pinned',   fingerprint }  — match, existing pin
 *   { status: 'new',      fingerprint }  — no prior pin, caller may pin
 *   { status: 'mismatch', fingerprint, expected } — active MITM or rotation
 */
export async function checkPin(peerId, remoteSpkiBytes) {
  const incoming = await computeFingerprint(remoteSpkiBytes);
  const pin = await getPin(peerId);
  if (!pin || !pin.fingerprint) return { status: 'new', fingerprint: incoming };
  if (pin.fingerprint === incoming) return { status: 'pinned', fingerprint: incoming };
  return { status: 'mismatch', fingerprint: incoming, expected: pin.fingerprint };
}

/** Convert a stored pin's pubSpkiB64 back into raw bytes. */
export function pubSpkiBytesFromPin(pin) {
  if (!pin || !pin.pubSpkiB64) return null;
  return base64ToBytes(pin.pubSpkiB64);
}

File: lib/core/peer_pins.dart
// Port of src/core/peerPins.js — TOFU pin store for remote identity keys.
//
// The first time a peer completes a signed (v3+) handshake we pin the SHA-256
// fingerprint of their identity pubkey. Every subsequent handshake must hash
// to the same fingerprint; a mismatch flags either legitimate key rotation
// (app reinstall) or an active MITM. We deliberately do not auto-overwrite —
// the user has to clear the pin before a different key is accepted.

import 'dart:typed_data';

import 'base64_helpers.dart';
import 'identity_key.dart' as identity_key;
import 'key_store.dart';

const String _pinPrefix = 'peer-pin-';
const String _keysTable = 'keys';

String _rowKey(String peerId) => '$_pinPrefix$peerId';

/// Stored pin snapshot. [pinnedAt] is a unix epoch ms.
class PeerPin {
  const PeerPin({
    required this.peerId,
    required this.pubSpkiB64,
    required this.fingerprint,
    required this.pinnedAt,
  });
  final String peerId;
  final String pubSpkiB64;
  final String fingerprint;
  final int pinnedAt;
}

Future<PeerPin?> getPin(String peerId) async {
  if (peerId.isEmpty) return null;
  final row = await keyStore().get(_keysTable, _rowKey(peerId));
  if (row == null) return null;
  final b64 = row['pubSpkiB64'];
  if (b64 is! String || b64.isEmpty) return null;
  return PeerPin(
    peerId: (row['peerId'] as String?) ?? peerId,
    pubSpkiB64: b64,
    fingerprint: (row['fingerprint'] as String?) ?? '',
    pinnedAt: (row['pinnedAt'] as num?)?.toInt() ?? 0,
  );
}

/// Pin [pubSpkiBytes] as the trusted identity for [peerId]. Returns the
/// computed fingerprint so callers can surface it in the UI immediately.
Future<({String fingerprint})> setPin(
    String peerId, List<int> pubSpkiBytes) async {
  if (peerId.isEmpty) {
    throw ArgumentError('setPin: peerId required');
  }
  final fingerprint = await identity_key.computeFingerprint(pubSpkiBytes);
  await keyStore().put(_keysTable, {
    'id': _rowKey(peerId),
    'peerId': peerId,
    'pubSpkiB64': bytesToBase64(Uint8List.fromList(pubSpkiBytes)),
    'fingerprint': fingerprint,
    'pinnedAt': DateTime.now().millisecondsSinceEpoch,
  });
  return (fingerprint: fingerprint);
}

Future<bool> deletePin(String peerId) async {
  if (peerId.isEmpty) return false;
  await keyStore().delete(_keysTable, _rowKey(peerId));
  return true;
}

/// Pin-check outcome. The JS side uses string literals — we keep the same
/// vocabulary so UI code can port one-to-one.
enum PinStatus { pinned, newPin, mismatch }

class PinCheck {
  const PinCheck({required this.status, required this.fingerprint, this.expected});
  final PinStatus status;
  final String fingerprint;
  final String? expected;
}

/// Verify a remote pubkey against the existing pin.
Future<PinCheck> checkPin(String peerId, List<int> remoteSpkiBytes) async {
  final incoming = await identity_key.computeFingerprint(remoteSpkiBytes);
  final pin = await getPin(peerId);
  if (pin == null || pin.fingerprint.isEmpty) {
    return PinCheck(status: PinStatus.newPin, fingerprint: incoming);
  }
  if (pin.fingerprint == incoming) {
    return PinCheck(status: PinStatus.pinned, fingerprint: incoming);
  }
  return PinCheck(
    status: PinStatus.mismatch,
    fingerprint: incoming,
    expected: pin.fingerprint,
  );
}

/// Convert a stored pin's base64 SPKI back into raw bytes.
Uint8List? pubSpkiBytesFromPin(PeerPin? pin) {
  if (pin == null || pin.pubSpkiB64.isEmpty) return null;
  return base64ToBytes(pin.pubSpkiB64);
}
